Add BIP for OP_TXHASH and OP_CHECKTXHASHVERIFY (#1500)
bitcoin/bips: bip-0346.md

> ```

Reviewer: ```suggestion
<pre>
```

  BIP: 346
  Layer: Consensus (soft fork)
  Title: OP_TXHASH and OP_CHECKTXHASHVERIFY
  Author: Steven Roose <steven@roose.io>, Brandon Black <freedom@reardencode.com>
  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0346
  Status: Draft
  Type: Standards Track
  Created: 2023-09-03
  License: BSD-3-Clause

> ```

Reviewer: ```suggestion
</pre>
```

# Abstract

This BIP proposes two new opcodes, `OP_CHECKTXHASHVERIFY`, to be activated
as a change to the semantics of `OP_NOP4` in legacy script, segwit and tapscript;

> and `OP_TXHASH`, to be activated as a change to the semantics of `OP_SUCCESS189`

Reviewer: I double checked, OP_SUCCESS appears to be unclaimed, OP_VAULT used `OP_SUCCESS187` and `OP_SUCCESS188`.

in tapscript only.

These opcodes provide a generalized method for introspecting certain details of
the spending transaction, which enables non-interactive enforcement of certain
properties of the transaction spending a certain UTXO.

The constructions specified in this BIP also open up the way for other
potential updates; see Motivation section for more details.

As mentioned in the Deployment section, alternatively, the
`OP_CHECKTXHASHVERIFY` semantics could be used to upgrade an existing deployment
of the `OP_CHECKTEMPLATEVERIFY` opcode. The opcode `OP_TXHASH` could then
potentially be renamed to `OP_TEMPLATEHASH`.

> # Summary

Reviewer: Should this section perhaps be called "Specification"?

## OP_CHECKTXHASHVERIFY

The first new opcode, `OP_CHECKTXHASHVERIFY`, redefines the `OP_NOP4` opcode

> (`0xb3`) as a soft fork upgrade. This opcode is active in both legacy script,
> segwitv0 p2wsh and tapscript contexts.

Reviewer: I count three different contexts, so maybe:

```suggestion
(`0xb3`) as a soft fork upgrade. This opcode is active in legacy script,
segwitv0 p2wsh, and tapscript contexts.
```

It has the following semantics:

* There is at least one element on the stack, fail otherwise.
* The element on the stack is at least 32 bytes long, NOP otherwise.
* The first 32 bytes are interpreted as the StackTxHash and the remaining
  suffix bytes specify the TxFieldSelector.

> * The TxFieldSelector valid, fail otherwise.

Reviewer: ```suggestion
* The TxFieldSelector is valid, fail otherwise.
```

> * The TxFieldSelector valid, fail otherwise.

Reviewer 2: Still digesting this proposal so apologies if bad question, but it might make sense for this opcode to trivially succeed if the tx field selector is not well defined by this BIP? This gives us flexibility in the future with extending how tx's are hashe? 🤷‍♂️

Author: That's not possible. We want to push something to the stack and going from nothing to the stack to something on the stack is a breaking change. Unfortunately it's not possible to have softfork-able hooks into this opcode.

* The TxHash of the transaction is calculated at the current input index using
  the TxFieldSelector.
* The StackTxHash equals the TxHash, fail otherwise.

## OP_TXHASH

The second new opcode, `OP_TXHASH`, redefines the `OP_SUCCESS189` tapscript

> opcode (`0xbd`) as a soft fork upgrade. This opcode is only active in tapscript
> context.

Reviewer: maybe "in tapscript" or "in the context of tapscript"?

```suggestion
opcode (`0xbd`) as a soft fork upgrade. This opcode is only active in tapscript.
```

It has the following semantics:

* There is at least one element on the stack, fail otherwise.
* The element is interpreted as the TxFieldSelector and is popped off the stack.
* The TxFieldSelector must be valid, fail otherwise.
* The 32-byte TxHash of the transaction at the current input index, calculated
  using the given TxFieldSelector is pushed onto the stack.

## TxFieldSelector

The TxFieldSelector has the following semantics. We will give a brief conceptual
summary, followed by a reference implementation of the CalculateTxHash function.

* There are two special cases for the TxFieldSelector:
  * the empty value, zero bytes long: it is set equal to `TXFS_SPECIAL_TEMPLATE`,
    the de-facto default value which means everything except the prevouts and the prevout
    scriptPubkeys and amounts.

    Special case `TXFS_SPECIAL_TEMPLATE` is 4 bytes long, as follows:

>     * 1: `TXFS_VERSION | TXFS_LOCKTIME | TXFS_CURRENT_INPUT_IDX`

Reviewer: How is the pipe symbol to be interpreted here?

    * 2: `TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_OUTPUTS_ALL`
    * 3: `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
    * 4: `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`

  * If the TxFieldSelector has exactly 1 byte, it has its 8 bits
    assigned as follows, from lowest to highest:
    * 2/1: Inputs
      * 00: `TXFS_INOUT_SELECTION_NONE`
      * 01: `TXFS_INOUT_SELECTION_CURRENT`
      * 11: `TXFS_INOUT_SELECTION_ALL`
    * 4/3: Outputs
      * 00: `TXFS_INOUT_SELECTION_NONE`
      * 01: `TXFS_INOUT_SELECTION_CURRENT`
      * 11: `TXFS_INOUT_SELECTION_ALL`
    * 5: `TXFS_INPUTS_PREVOUTS`
    * 6: `TXFS_INPUTS_PREV_SCRIPTPUBKEYS | TXFS_INPUTS_PREV_VALUES`
    * 7: `TXFS_CURRENT_INPUT_CONTROL_BLOCK | TXFS_CURRENT_INPUT_SPENTSCRIPT | TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS`
    * 8: `TXFS_CURRENT_INPUT_IDX`

    Additionally, it includes `TXFS_VERSION | TXFS_LOCKTIME | TXFS_CONTROL | TXFS_CURRENT_INPUT_TAPROOT_ANNEX | TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_OUTPUTS_ALL`

    These 1-byte selections allow the TxFieldSelector to emulate current

>     signature hashing modes and those defined in BIP-118:

Reviewer: ```suggestion
    signature hashing modes and those defined in [BIP-118: SIGHASH_ANYPREVOUT for Taproot Scripts](bip-0118.mediawiki):
```

| BIP-341/118 sighash type     | 1-byte TxFieldSelector |
| :--------------------------- | :--------------------- |
| `ALL`                        | `0b11111111`           |
| `SINGLE`                     | `0b11110111`           |
| `NONE`                       | `0b11110011`           |
| `ALL|ANYONECANPAY`           | `0b11111101`           |
| `SINGLE|ANYONECANPAY`        | `0b11110101`           |
| `NONE|ANYONECANPAY`          | `0b11110001`           |
| `ALL|ANYPREVOUT`             | `0b11101101`           |
| `SINGLE|ANYPREVOUT`          | `0b11100101`           |
| `NONE|ANYPREVOUT`            | `0b11100001`           |
| `ALL|ANYPREVOUTANYSCRIPT`    | `0b11001101`           |
| `SINGLE|ANYPREVOUTANYSCRIPT` | `0b11000101`           |
| `NONE|ANYPREVOUTANYSCRIPT`   | `0b11000001`           |

> * The first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:

Reviewer: It seems to me that we are now in the case where the TxFieldSelector is longer than 1 byte. It would perhaps be helpful to explicitly communicate that:

```suggestion
* If the TxFieldSelector is longer than 1 byte, the first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:
```

  * 1: version (`TXFS_VERSION`)
  * 2: locktime (`TXFS_LOCKTIME`)
  * 3: current input index (`TXFS_CURRENT_INPUT_IDX`)
  * 4: current input control block (or empty) (`TXFS_CURRENT_INPUT_CONTROL_BLOCK`)
  * 5: current input spent script (i.e. witness script or tapscript) (`TXFS_CURRENT_INPUT_SPENTSCRIPT`)
  * 6: current script last `OP_CODESEPARATOR` position (or 0xffffffff)
    (`TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS`)
  * 7: current input annex (or empty) (`TXFS_CURRENT_INPUT_TAPROOT_ANNEX`)
  * 8: `TXFS_CONTROL` (i.e. include TxFieldSelector into hash)

* The highest bit of the first byte (`TXFS_CONTROL`), we will call the
  "control bit", and it can be used to control the behavior of the opcode. For
  `OP_TXHASH` and `OP_CHECKTXHASHVERIFY`, the control bit is used to determine
  whether the TxFieldSelector itself has to be included in the resulting hash.
  (For potential other uses of the TxFieldSelector (like a hypothetical
  `OP_TX`), this bit can be repurposed.)

* The second byte will be used to indicate fields from the inputs and outputs.

>   If there is only a single byte present, no information from the inputs and

Reviewer: I’m probably missing something, but the option for the TxFieldSelector to only have a single byte here appears to be at odds with the special case above in line 83 "If the TxFieldSelector has exactly 1 byte, […]"

  outputs will be committed. Otherwise, of the second byte, the 8 bits are
  assigned the following variables, from lowest to highest:
  * Specifying which fields of the inputs will be selected:
    * 1: prevouts (`TXFS_INPUTS_PREVOUTS`)
    * 2: sequences (`TXFS_INPUTS_SEQUENCES`)
    * 3: scriptSigs (`TXFS_INPUTS_SCRIPTSIGS`)
    * 4: prevout scriptPubkeys (`TXFS_INPUTS_PREV_SCRIPTPUBKEYS`)
    * 5: prevout values (`TXFS_INPUTS_PREV_VALUES`)
    * 6: taproot annexes (`TXFS_INPUTS_TAPROOT_ANNEXES`)

  * Specifying which fields of the outputs will be selected:
    * 7: scriptPubkeys (`TXFS_OUTPUTS_SCRIPTPUBKEYS`)
    * 8: values (`TXFS_OUTPUTS_VALUES`)

* We define as follows:
  * `TXFS_ALL = TXFS_VERSION | TXFS_LOCKTIME | TXFS_CURRENT_INPUT_IDX | TXFS_CURRENT_INPUT_CONTROL_BLOCK | TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS | TXFS_CONTROL`
  * `TXFS_INPUTS_ALL = TXFS_INPUTS_PREVOUTS | TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_INPUTS_PREV_SCRIPTPUBKEYS | TXFS_INPUTS_PREV_VALUES | TXFS_INPUTS_TAPROOT_ANNEXES`

>   * `TXFS_OUTPUTS_ALL = TXFS_OUTPUTS_SCRIPTPUBKEYS | TXFS_OUTPUTS_VALUES`

Reviewer: Please explain the semantics of these lines. Does the pipe represent string concatenation, or some other form of concatenation here?

## Resource limits

* For legacy scripts and segwit, we don't add any extra resource limitations,
  with the argumentation that `OP_CHECKTXHASHVERIFY` already requires the user
  to provide at least 32 bytes of extra transaction size, either in the input
  scriptSig, or the witness. Additional more complex hashes require additional
  witness bytes. Given that `OP_CAT` is not available in this context, if a
  malicious user tries to increase the number of TransactionHashes being
  calculated by using opcodes like `OP_DUP`, the TxFieldSelector for all these
  calculations is identical, so the calculation can be cached within the same
  transaction.

* For tapscript, primarily motivated by the cheaper opcode `OP_TXHASH` (it
  doesn't require an additional 32 witness bytes be provided) and the potential
  future addition of byte manipulation opcodes like `OP_CAT`, an additional
  cost is specified per TransactionHash execution. Using the same validation
  budget ("sigops budget") introduced in BIP-0342, each TransactionHash
  decreases the validation budget by 25. If this brings the budget below zero,
  the script fails immediately.<br>The following considerations should be made:

>   * All fields that can be of arbitrary size are cachable as TransactionHash

Reviewer: ```suggestion
  * All fields that can be of arbitrary size are cacheable as TransactionHash
```

    always hashes their hashed values.
  * In "individual mode", a user can at most commit 32 inputs or outputs,
    which we don't consider excessive for potential repeated use.
  * In "leading mode", a caching strategy can be used where the SHA256 context
    is stored every N in/outputs so that multiple executions of the
    TransactionHash function can use the caches and only have to hash an
    additional N-1 items at most.

# Motivation

This BIP specifies a basic transaction introspection primitive that is useful
to either reduce interactivity in multi-user protocols or to enforce some basic
constraints on transactions.

Additionally, the constructions specified in this BIP can lay the groundwork for
some potential future upgrades:
* The TxFieldSelector construction would work well with a hypothetical opcode
  `OP_TX` that allows for directly introspecting the transaction by putting the
  fields selected on the stack instead of hashing them together.

> * The TransactionHash obtained by `OP_TXHASH` can be combined with a
>   hypothetical opcode `OP_CHECKSIGFROMSTACK` to effectively create an

Reviewer: You mean like the one proposed in BIP 348? ;)

```suggestion
* The TransactionHash obtained by `OP_TXHASH` can be combined with the
  proposed opcode `OP_CHECKSIGFROMSTACK` (see [BIP-348](bip-0348.md)) to effectively create an
```

  incredibly flexible signature hash, which would enable constructions like
  `SIGHASH_ANYPREVOUT`.

## Comparing with some alternative proposals

> * This proposal strictly generalizes BIP-119's `OP_CHECKTEMPLATEVERIFY`, as the

Reviewer: ```suggestion
* This proposal strictly generalizes [BIP-119](bip-0119.mediawiki)'s `OP_CHECKTEMPLATEVERIFY`, as the
```

  default mode of our TxFieldSelector is semantically the same (though not
  byte-for-byte identical) as what `OP_CTV` accomplishes, without costing any
  additional bytes. Additionally, using `OP_CHECKTXHASHVERIFY` allows for more
  flexibility which can help in the case for
  * enabling adding fees to a transaction without breaking a multi-tx protocol;
  * multi-user protocols where users are only concerned about their own inputs and outputs.

* Constructions like `OP_IN_OUT_VALUE` used with `OP_EQUALVERIFY` can be
  emulated by two `OP_TXHASH` instances by using the TxFieldSelector to select
  a single input value first and a single output value second and enforcing
  equality on the hashes. Neither of these alternatives can be used to enforce
  small value differencials without the availability of 64-bit arithmetic in
  Script.

* Like mentioned above, `SIGHASH_ANYPREVOUT` can be emulated using `OP_TXHASH`
  when combined with `OP_CHECKSIGFROMSTACK`: 
  `<txfs> OP_TXHASH <pubkey> OP_CHECKSIGFROMSTACK` effectively emulates `SIGHASH_ANYPREVOUT`.

> # Detailed Specification

Reviewer: This seems to be talking about the Reference Implementation, not a Specification?

A reference implementation in Rust is provided attached as part of this BIP
together with a JSON file of test vectors generated using the reference
implementation.

> # Backwards Compatibility

Reviewer: This might also be a good place to mention the interactions with BIP 118, BIP 119, BIP 347, and BIP 348.

  License: BSD-3-Clause
```

# Abstract

This BIP proposes a new opcode `OP_TXHASH`, to be activated as a change to the
semantics of `OP_SUCCESS189` in tapscript contexts.

This opcode provides a generalized method for introspecting certain details of
the spending transaction, which enables non-interactive enforcement of certain
properties of the transaction spending a certain UTXO.

Together with an opcode like `OP_CHECKSIGFROMSTACK`, this opcode effectively
provides a fully generalized signature hash construction, fully supporting
all existing SIGHASH flags, the proposed sighash flags from
[BIP-118](bip-0118.mediawiki) (`SIGHASH_ANYPREVOUT`) and many other new signature hash
combinations.

The constructions specified in this BIP also open up the way for other
potential updates; see Motivation section for more details.

As mentioned in the Deployment section, alternatively, the `OP_TXHASH` semantics
could be used to upgrade an existing deployment of the
`OP_CHECKTEMPLATEVERIFY` opcode.

# Specification

## OP_TXHASH

`OP_TXHASH` redefines the `OP_SUCCESS189` tapscript opcode (`0xbd`) as a soft
fork upgrade. This opcode is only active in tapscript contexts. Note that

> `OP_SUCCESS187` and `OP_SUCCESS188` are used by BIP-345 (OP_VAULT), making

Reviewer: Nit: OP_VAULT was withdrawn, so it no longer lays claim to those. However, BIP443: OP_CHECKCONTRACTVERIFY now claims OP_SUCCESS187.
